data/basics/teal.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package basics
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+)
+
+// DeltaAction is an enum of actions that may be performed when applying a
+// delta to a TEAL key/value store
+type DeltaAction uint64
+
+const (
+	// SetUintAction indicates that a Uint should be stored at a key
+	SetUintAction DeltaAction = 1
+
+	// SetBytesAction indicates that a TEAL byte slice should be stored at a key
+	SetBytesAction DeltaAction = 2
+
+	// DeleteAction indicates that the value for a particular key should be deleted
+	DeleteAction DeltaAction = 3
+)
+
+// ValueDelta links a DeltaAction with a value to be set
+type ValueDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Action DeltaAction `codec:"at"`
+	Bytes  string      `codec:"bs"`
+	Uint   uint64      `codec:"ui"`
+}
+
+// ToTealValue converts a ValueDelta into a TealValue if possible, and returns
+// ok = false with a value of 0 TealUint if the conversion is not possible.
+func (vd *ValueDelta) ToTealValue() (value TealValue, ok bool) {
+	switch vd.Action {
+	case SetBytesAction:
+		value.Type = TealBytesType
+		value.Bytes = vd.Bytes
+		ok = true
+	case SetUintAction:
+		value.Type = TealUintType
+		value.Uint = vd.Uint
+		ok = true
+	case DeleteAction:
+		value.Type = TealUintType
+		ok = false
+	default:
+		value.Type = TealUintType
+		ok = false
+	}
+	return value, ok
+}
+
+// StateDelta is a map from key/value store keys to ValueDeltas, indicating
+// what should happen for that key
+//msgp:allocbound StateDelta config.MaxStateDeltaKeys
+type StateDelta map[string]ValueDelta
+
+// Equal checks whether two StateDeltas are equal. We don't check for nilness
+// equality because an empty map will encode/decode as nil. So if our generated
+// map is empty but not nil, we want to equal a decoded nil off the wire.
+func (sd StateDelta) Equal(o StateDelta) bool {
+	// Lengths should be the same
+	if len(sd) != len(o) {
+		return false
+	}
+	// All keys and deltas should be the same
+	for k, v := range sd {
+		if o[k] != v {
```

Comment:
this would work, but not for the reason you've mentioned.
If v happen to have the default value and o[k] doesn't exist, the function might still return true.
however, this could not happen in reality, since the DeltaAction is always non-zero.

Reply:
oh wow, good catch! that's terrible. fixing.

Reply:
pushed a fix, thanks again!